### PR TITLE
Improve historical Node data

### PR DESCRIPTION
**Pull request by zloirock**

### CODE REVIEW

**Comment by ljharb** [comment]
How are all these verified? node older than 0.6 won’t compile on any hardware i have access to.

File: environments.json [hunk 1]
```diff
@@ -2554,15 +2837,7 @@
...
-    "obsolete": true,
-    "test_suites": [
-      "es5",
-      "es6",
-      "es2016plus",
-      "esnext",
-      "esintl",
-      "non-standard"
-    ]
+    "obsolete": true
```
Comment (chicoxyzzy) [inline environments.json:2675]: should we change it to `"very"` here?

**Comment by zloirock** [comment]
@ljharb by V8 versions.

File: environments.json (hunk 1 above)
Comment (zloirock) [reply environments.json:2675]: I'm not sure about 0.10+. Some users prefer node.green fork since they can't find the required Node version here.

**Review by ljharb — CHANGES_REQUESTED** [review]
node floats v8 patches all the time, even historically; these changes need manual verification with the actual node versions, not just comparing v8 versions.

**Comment by zloirock** [comment]
Feel free to check it. I haven't required hardware, you - too, but the version from the current PR definitely better than we have now (difference in 10 Chrome / V8 versions with the real result in 0.10, 0.12, iojs1, etc).

**Comment by zloirock** [comment]
But if you wanna check it, please, do it out of the scope of this PR which will not wait for it.

**Comment by ljharb** [comment]
Yes, it will wait for it; this table is for manually verified data, and things don’t land on this repo when a collaborator blocks it, as you well know.

Please feel free to remove changes you can’t manually verify (anything below node 0.8 is likely to be tricky, but the rest is trivial with nvm) and I’ll be on board.

**Comment by zloirock** [comment]
If you wanna do it - do in a separate PR, please. This PR will be merged anyway, even without your destructive "review". If you want to set any rule for this repository, please do at least something useful for the project. For example, the first PR for the 2 years with validation of this result.

**Comment by ljharb** [comment]
Validating PRs alone is useful. If this PR is merged anyways, I’ll just revert it, and then we’ll have to get the repo owner involved - which won’t likely be good for anybody.

Nothing lands in any github repo unless collaborators don’t object. This isn’t my rule, it’s just how it’s always worked - especially here, but also anywhere. If you’ll recall I’ve had some changes that you’ve blocked, and I’ve never done you the discourtesy of merging through your block, conceptual or concrete. Please treat me with the same respect.

**Comment by zloirock** [comment]
If you will revert it - I'll revert it back. If you will show where I can read this "rule" about verification critically obsolete engines on hardware - I'll think about it, otherwise - it's your work. Please do not engage in destructive activities, interfering with improving the project.

**Comment by ljharb** [comment]
Landing data unverified by a human is destructive.

**Comment by zloirock** [comment]
...and now we store data with

> difference in 10 Chrome / V8 versions with the real result in 0.10, 0.12, iojs1, etc

One more time - if you wanna verify it on the real hardware - do it, I can't do it, no rule forces me to do this and it's not my or users problem.

**Comment by chicoxyzzy** [comment]
I agree with @ljharb that adding unverified results is not what we were doing historically. +1 for removing unverifiable data.

**Comment by zloirock** [comment]
Ok.

**Comment by ljharb** [comment]
To be clear, verification of everything prior to merge is what i asked for, not just removing the unverifiable ones. Did you try out every node version covered by this PR yourself?

**Comment by zloirock** [comment]
Feel free to check them all if you need it, it's not my problem. You wrote that you can't check 0.6-, so now it's work for you. I checked some of them selectively.

**Comment by ljharb** [comment]
The burden of verifying a change is on whoever wants to land it; you don’t land things without tests and say “not my problem, write the tests yourself”.

**Comment by zloirock** [comment]
If you don't wanna improve this project - please, don't cause problems for those who want to do this.

**Comment by ljharb** [comment]
You haven’t improved it - reducing the data quality makes it worse. Leaving it frozen in time forever is more of an improvement than ruining the data quality.

**Comment by zloirock** [comment]
Apparently you can't read.

> difference in 10 Chrome / V8 versions with the real result in 0.10, 0.12, iojs1, etc

Perfect data quality -)

**Comment by kangax** [comment]
@zloirock @ljharb folks, you both reached out to me personally to resolve or comment on this. It's sad to see this come to this. Clearly, you both have strong opinions and I'm sure you both have reasons for doing things certain way. It's not my place to "chose" one of you or ban someone. It's not my personal project (anymore), I haven't been contributing to it in ages, and you've both done incredible work over the years — either with direct commits or reviewing and commenting on things.

So, how do we resolve this?

In most of the places I've worked at, the rule for merging is simple — a PR needs at least one thumb from another person. This usually avoids situations like this and shifts responsibility onto the majority of votes. A single person should not be able to make (potentially controversial) decisions.

I think this would be the most democratic and reasonable solution BUT I don't know if we have enough contributors (@chicoxyzzy seems to be the only other person) for this to work effectively? In absence of 3+ contributors, this could just lead to a painful drag where no one stamps and merges things (or blocks the other).

Let me know what you think.

**Comment by ljharb** [comment]
@kangax the unspoken rule we've tended to follow over the years is that two collaborators' approvals (counting the PR author's implicit one) is sufficient to merge, but that a block from any collab may never be overridden. This mirrors many other open source projects where 100% approval is not required, but 0% rejection is. (Many projects also make sure that a PR that is potentially controversial "sits" for awhile, to give the most people the best opportunity to review - iow, it wouldn't be acting in good faith to have two collaborators "team up" to rush and get PRs in before another has a chance to voice an objection)

I doubt we will have any issues with 2 of the 3 of us gaining approvals, and if we do, we can certainly come to an agreement on some kind of "waiting period" - but I don't think it's ever going to be a good idea to override objections of other collaborators.

**Comment by zloirock** [comment]
@kangax that makes sense. But...

Requirements of approvement that no one wanna check for some days (like #1518) is a bad idea.

Unreasonable block like in this case - it's bull shit. Actions of @ljharb are sabotage since it's adding fixes for Node 0.12 / 0.10 compat data mapping which currently is completely broken, still shown in the table and used by Babel. If he don't wanna improve `compat-table`, he should not interfere with those who do this. 2 approvements should beat a block.

**Comment by ljharb** [comment]
Who decides what's "unreasonable" though? If specific data for node 0.10 and 0.12 is broken, then a PR that fixes that - including manually verifying the new results - is a great PR to add!

However, just like bugfixes without regression tests are often not landed until those tests are written (whether that takes hours, days, or months), data that's allegedly an improvement but lacks manual verification can wait too until the proper checking is done.

**Comment by zloirock** [comment]
@ljharb feel free to check it. Current `node` script does not work in ancient versions, I have no access to some ancient Node versions - but I'm sure in the results. If you are not sure - you can rewrite the script, check it and update the result. Forcing me to do this as part of this PR is stupid.

**Comment by ljharb** [comment]
The attitude of "feel free to do the extra work yourself" is the one I'm taking issue with. The burden of doing the work is on the PR author (the person making the change), not anyone else. If you're interested in your change landing, it's *on you* to satisfy the other decisionmakers on the project.

I'm not aware of any multiple-contributor project - open source or corporate - that works differently.

**Comment by zloirock** [comment]
Sorry, but you have not been adding anything to this project for many years - only periodically review PRs. It’s not for you to force me to rewrite this script and manually test hundreds of Node versions and spend dozen of hours for that.

Maybe also I should buy hardware where I should test them?

Maybe also I should manually test all Opera Mobile or Samsung Internet version and buy hardware for that? -) The same Chromium / V8 mapping like here. About built-ins related differences in Node with Chromium I know from the changelogs.

**Comment by ljharb** [comment]
Reviewing PRs *is* adding value to projects; code is not the only (or even the most important) kind of contribution.

I'm not sure what your development environment is that means you can't test various node versions - but yes, if you're incapable of testing an environment and also can't find someone willing to volunteer, then you are not able to reliably make changes to that environment's data. I"m not asking you to rewrite any script (nor do I see any such script; perhaps adding that to the PR would encourage volunteers to help) - but you are asking the rest of us to accept untested data.

**Comment by zloirock** [comment]
You wrote that you can test Node >=0.6 - I asked you about it - but it seems you can only "review" PRs and enforce others to do something -) Yes, reviewing has a value - but, in this case, without adding something new - very small value.

**Comment by ljharb** [comment]
That i have the capability doesn’t mean i have the time.

Maintaining invariants is a much larger value than adding something new.

**Comment by zloirock** [comment]
Maintaining invariants is a much larger value than just reviewing PRs without deep knowledge of the project. You are "maintain" too many projects - but how many of them you know enough good?

At this moment for this project adding new content required much more than reviewing PRs.

If you don't have the time - it does not mean that I have time for that. Even in the scope of this project, I can spend time much more productive. However, this fix required and ASAP.

**Comment by ljharb** [comment]
I don’t see why - nothing in the PR indicates a concrete bug or problem being fixed. Note the “no description given” in the part of the PR where you’re supposed to explain why the change is needed and important? https://chris.beams.io/posts/git-commit/ may be helpful.

@kangax I’m afraid that it’s unlikely either of us will change our stance; i think the real question you will need to dictate is whether a PR can be merged through a block, or not. The rest is something that we can hopefully figure out with ourselves with the hope that we’re all operating in good faith.

**Comment by zloirock** [comment]
I was hoping that you'll take a look at least on commits names and them [content](https://github.com/kangax/compat-table/pull/1556/commits/72aed7cf610ed95d594b615474646555fa490ba2) 🤦

**Comment by atassis** [comment]
Guys (@ljharb @zloirock) , it is clean that you both can't agree with each other, so instead of shitting on each other (pardon my french), u'ld better follow the @kangax 's advice to let the majority decide. If you don't follow his suggestion- why did you even ask his opinion as he mentioned in his reply? I am none from nowhere, but think that it is not ethical to act as you do in an open source project. Please, respect each other by using common discussion rules, one of which is the mentioned advice.

**Comment by zloirock** [comment]
@kangax until this sabotage will not be stopped and we will not have a way to resist it - I'll stop any contributions to `compat-table`, maybe excepting adding `core-js`-related data and reviewing PRs. Let @ljharb add new content.

**Comment by chicoxyzzy** [comment]
It became really hard to contribute to this project. It's almost impossible to rebase any changes against constantly changing gh-pages branch.

**Comment by chicoxyzzy** [comment]
https://github.com/kangax/compat-table/issues/1564

**Pull request merged**